### PR TITLE
added XML namespace support

### PR DESCRIPTION
**Pull request by ipa-mdl**
This patch relaxes the checking of unknown attributes in `package.xml` to support [XML namespaces](https://www.w3.org/TR/xml-names/).
`xml.dom.minidom` already checks the validity of the prefixes (see test).

Use cases:
* provide schema in `xsi:schemaLocation` (not all parsers support [`xml-model`](https://www.w3.org/TR/xml-model/#dt-xml-model))
* XMI data in `<export>`
* tag meta-data annotations, e.g. for generators or cross-linking

### CODE REVIEW

**Comment by dirk-thomas** [comment]
Does this require and update to REP [127](http://www.ros.org/reps/rep-0127.html) and [141](http://www.ros.org/reps/rep-0141.html)?

Do manifest files using these attributes validate using the current schemas (format [1](http://download.ros.org/schema/package_format1.xsd) and [2](http://download.ros.org/schema/package_format2.xsd))?

**Comment by ipa-mdl** [comment]
> Does this require and update to REP 127 and 141?

Not necessarily, namespaces are more or less an orthogonal XML feature.
There is a section on [XML restrictions](http://www.ros.org/reps/rep-0127.html#xml-syntax-restrictions), but this does not conflict with the usage of namespaces.
This could be stated explicitly, but I don't think that this is needed.

> Do manifest files using these attributes validate using the current schemas (format 1 and 2)?

Yes, I have tested it with a patched version of the [example in the wiki](http://wiki.ros.org/catkin/package.xml#Required_Tags):
```xml
<package format="2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <name>foo_core</name>
  <version xsi:type="VersionType">1.2.4</version>
  <description>
  This package provides foo capability.
  </description>
  <maintainer email="ivana@osrf.org">Ivana Bildbotz</maintainer>
  <license>BSD</license>
</package>
```
[example.xml.txt](https://github.com/ros-infrastructure/catkin_pkg/files/1377256/example.xml.txt)

```
$ xmllint --noout --schema  http://download.ros.org/schema/package_format2.xsd example.xml.txt 
example.xml.txt validates
```

**Comment by ipa-mdl** [comment]
It might make sense to add this example as an addtional unit test.

**Comment by dirk-thomas** [comment]
A parser which can read manifests currently specified by the REP might choke on `xsi:type="VersionType"` since it doesn't expect it to be allowed. I am always hesitant to change a spec like this after the fact. There *might* be code out there which break - there is just no way of knowing. We could either amend the REPs 127 and 140 to explicitly mention XML namespaces or only allow them in an upcoming format 3.

@ros-infrastructure/ros_team What do others think about this?

**Comment by ipa-mdl** [comment]
The example was just given for the validation, `xsi:type="VersionType"` does not make much sense.
I'd like to use these things in the `<export>` section, but the namespace might get defined in the root element in order to be references later.
It should not break any parser.

**Comment by dirk-thomas** [comment]
> It should not break any parser.

"Should". Just to make an example: someone might have implemented a custom parser to fit on a micro controller and it will simply fail if the root element has an attribute not specified in the REP. I just want to point out that we are aware of the possibility and a change in the REP would have that risk of breakage.

**Comment by wjwwood** [comment]
tl;dr +0, I don't see why we need it exactly, but the potential harm I think is unlikely to occur.

I don't see the impetus for allowing the XML namespaces (I'm sure there are use cases, but I haven't seen a lot of people asking for it). But on the other hand I think it's unlikely very many people are using a non-compliant XML parser to interpret the manifest files, and if they are, we did declare to be using XML, so it would be a bad assumption that we don't use a feature of the specification. And in that case I think it's fine for custom parsers to be broken and need to switch to a compliant parser or update their "custom" parser to allow or ignore these attributes.

**Comment by dirk-thomas** [comment]
Sounds good to me.

Thank you for the patch and thanks for the discussion / feedback.
